### PR TITLE
#159987732 Users can get their reading stats

### PR DESCRIPTION
**Pull request by emmaadesile**
### What does this PR do?
Users should be able to get their reading stats.

### Description of tasks to be completed
- User's userId is logged to the `ArticleView` model when they view an article
- Users can get their reading stats.
- Reading stats include the following
  - The articles a user has read
  - The articles that a user has liked/disliked
  - The the most read category

### How can this be manually tested
- After cloning the app
- run `npm install` to install all the dependencies
- install and setup Postgresql
- create a `.env` file using the `.envsample` file
- on Postman, log in as a user and view an article at `/api/articles/:article-slug`
- Goto `/api/user-reading-stats` to view the reading stats of the user

### Screenshots
reading-stats

### Relevant Pivotal Tracker stories
[159987732](https://www.pivotaltracker.com/story/show/159987732)

### CODE REVIEW

File: server/controllers/article.js [hunk 1]
```diff
@@ -211,10 +210,32 @@ class ArticleController {
...
+        const articleId = article.id;
+        const { userId } = req;
+        ArticleView.findOne({
+          where: { userId, articleId },
+        })
+          .then((userView) => {
+            if (!userView) {
+              ArticleView.create({ articleId, userId })
+                .then(() => res.status(201).json())
+                .catch(next);
```
Comment (CEOehis) [inline server/controllers/article.js]: The name of this method is `getSpecific`, I think it is not really descriptive. The method name should describe what it is getting. Also in lines`228-230` a resource is created which doesn't seem to tally with what I think the method should be doing.

I think you should have a separate method that creates an articleView.

File: server/controllers/readingStatsController.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+                      numberOfArticlesRead: articleCount.length,
+                      articleReactions: reaction.length === 0 ? 'You have not liked/disliked any article' : reaction,
+                      mostReadCategory
+                    });
+                  })
+                  .catch(error => res.status(400).send(error));
+              })
+              .catch(error => error);
+          })
+          .catch(error => error);
```
Comment (CEOehis) [inline server/controllers/readingStatsController.js]: It is good that you have caught this error, but just returning the error without handling it is not as user experience friendly as handling it. You can instead, call next with the error and let the general error handler defined in the apps `index.js` handle it or handle the errors as you did in line 76.

This also applies to line 72.

File: server/utils/modeValue.js [hunk 3]
```diff
@@ -0,0 +1,6 @@
+const mode = words => (
+  words.sort((a, b) => words.filter(v => v === a).length
+  - words.filter(v => v === b).length).pop()
+);
+
+export default mode;
```
Comment (CEOehis) [inline server/utils/modeValue.js]: Nice work on writing this utility method using functional Javascript style. I think you can make readers of your code understand what the utility does by using  JSDoc comments and also using more descriptive variable names for the variables in your array methods.

**Review by CEOehis — CHANGES_REQUESTED** [review]

File: server/controllers/readingStatsController.js (hunk 2 above)
Comment (emmaadesile) [reply server/controllers/readingStatsController.js]: Alright. I'll fix it now.

File: server/controllers/article.js (hunk 1 above)
Comment (emmaadesile) [reply server/controllers/article.js]: The name `getSpecific` was the work of @ozezzy, but I'll fix it now.

File: server/utils/modeValue.js (hunk 3 above)
Comment (emmaadesile) [reply server/utils/modeValue.js]: Good one. Duly noted. I'll make the changes now

File: server/controllers/article.js (hunk 1 above)
Comment (emmaadesile) [reply server/controllers/article.js]: Will work on it now

File: server/tests/controllers/category.test.js [hunk 4]
```diff
@@ -494,5 +472,22 @@ describe('Categorizes articles', () => {
...
+      chai.request(app)
+        .get('/api/article-categories/Technology/articles')
+        .set('Accept', 'application/json')
+        .set('Authorization', `Bearer ${userToken}`)
+        .set('Content-Type', 'application/json')
+        .end((req, res) => {
+          expect(res).to.have.status(200);
+          expect(res.body.status).to.equal('success');
+          expect(res.body).to.have.property('category');
+          expect(res.body.category).to.have.property('createdAt');
```
Comment (emekafredy) [inline server/tests/controllers/category.test.js]: It is always better to use object destructuring(lines 485 - 488). Also applies for lines 72 - 74 of your readingstats test file, lines 89 & 90.

**Review by emekafredy — CHANGES_REQUESTED** [review]
Please, also update the swagger doc with the new endpoint(s) you defined.

File: server/tests/controllers/category.test.js (hunk 4 above)
Comment (emmaadesile) [reply server/tests/controllers/category.test.js]: I'll do that now.

File: server/controllers/categoryController.js [hunk 5]
```diff
@@ -38,7 +39,7 @@ class CategoryController {
    */
   static createCategory(req, res) {
     const name = req.body.name.trim();
-    const newCategory = `${name.substr(0, 1).toUpperCase()}${name.slice(1).toLowerCase()}`;
+    const newCategory = sanitizeString(name);
     if (newCategory === '') {
```
Comment (aknwosu) [inline server/controllers/categoryController.js]: if (!newCategory) {

File: server/controllers/readingStatsController.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                  .then((categoryNames) => {
+                    let mostReadCategory;
+                    if (categoryNames.length === 0) {
+                      mostReadCategory = 'Articles have not been added to any category';
+                    }
+                    const names = categoryNames.map(category => category.name);
+                    mostReadCategory = mode(names);
+                    return res.status(200).json({
+                      status: 'success',
+                      articlesRead: articles.length === 0 ? 'You have not read any article' : articles,
```
Comment (aknwosu) [inline server/controllers/readingStatsController.js]: No articles found

File: server/controllers/readingStatsController.js [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+                      numberOfArticlesRead: articleCount.length,
+                      articleReactions: reaction.length === 0 ? 'You have not liked/disliked any article' : reaction,
+                      mostReadCategory
```
Comment (aknwosu) [inline server/controllers/readingStatsController.js]: No reactions found

File: server/controllers/readingStatsController.js [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  /**
+   * @description Get all categories
+   * @param  {object} req body of the user's request
+   * @param  {function} res response from the server
+   * @param  {function} next response from the server
+   * @returns {object} The body of the response message
+   * @param  {string} articleId body of the user's request
+   * @memberof ReadingStatsController
+   */
+  static saveArticleView(req, res, next) {
```
Comment (aknwosu) [inline server/controllers/readingStatsController.js]: postArticleViewHistory

File: server/models/articleview.js [hunk 9]
```diff
@@ -0,0 +1,13 @@
+module.exports = (sequelize, DataTypes) => {
+  const ArticleView = sequelize.define('ArticleView', {
```
Comment (aknwosu) [inline server/models/articleview.js]: articleViewHistory

**Review by aknwosu — CHANGES_REQUESTED** [review]

**Pull request merged**